### PR TITLE
Internal cell comm fault

### PR DESCRIPTION
**Pull request by Sabramz**
Added internal cell communication fault codes to functions in the segment interface that get voltage and thermistor data. Also changed the return values of the functions to a BMSfault_t in both the h and c files.

### CODE REVIEW

File: Core/Src/segment.c [hunk 1]
```diff
@@ -44,6 +44,12 @@ void SegmentInterface::retrieveSegmentData(ChipData_t databuf[NUM_CHIPS])
     memcpy(previous_data, segment_data, sizeof(ChipData_t)*NUM_CHIPS);
 
     segment_data = nullptr;
+
+    if (voltage_error) {
```
Comment (dyldonahue) [inline Core/Src/segment.c:48]: so this doesnt work perfectly. What if there is a therm error and a voltage error? you will miss the therm error. Our faults work in a way that you can use |= operator. This allows u to have a value that shows both faults. ie if voltage error is represented by bit 6, and therm error is bit 7, you can do like Val |= voltage error and do val |= therm error and the value wqould be like 1100 0000 if they both faulted. This is how u should return these values

**Review by dyldonahue — CHANGES_REQUESTED** [review]
As well as other comment, this fault isnt actually handled, just passed upstream. Look at the fault table in stateMachine and see what needs to be done. This one is a little weird, because the fault is being calculated downstream, where most of our faults are calculated in statemachine. Handle this by setting the parameter to check for in statemachine to be non zero. Meaning, If (return value from the pull sgement data that should have both therm and voltaghe fault) is not equal to zero, then fault. IDC about the timer length for now, but youll need to create a timer for it to.

File: Core/Inc/bmsConfig.h [hunk 2]
```diff
@@ -39,15 +39,16 @@
...
+#define OVER_CURR_TIME          1500 //todo adjust these based on testing and/or counter values
+#define PRE_OVER_CURR_TIME      1000
+#define OVER_CHG_CURR_TIME      1000
+#define UNDER_VOLT_TIME         15000
+#define PRE_UNDER_VOLT_TIME     12000
+#define OVER_VOLT_TIME          15000
+#define LOW_CELL_TIME           15000
+#define HIGH_TEMP_TIME          60000
+#define INTERN_CELL_COMM_TIME   1000
+#define CURR_ERR_MARG           50       // in A * 10
```
Comment (Sabramz) [inline Core/Inc/bmsConfig.h:51]: This is just adjusting formatting and adding INTERN_CELL_COMM_TIME

File: Core/Inc/segment.h [hunk 3]
```diff
@@ -10,6 +10,8 @@
 #define VOLTAGE_WAIT_TIME   100 //ms
 #define THERM_AVG           15 // Number of values to average
 
+#define CELL_COMM_FAULT_STATUS 0 
+
```
Comment (Sabramz) [inline Core/Inc/segment.h:14]: Variable for passing errors between segment and state machine

File: Core/Inc/segment.h [hunk 4]
```diff
@@ -58,9 +60,9 @@ class SegmentInterface
 
         void SelectTherm(uint8_t therm);
 
-        FaultStatus_t pullThermistors();
+        BMSFault_t pullThermistors();
 
-        FaultStatus_t pullVoltages();
+        BMSFault_t pullVoltages();
```
Comment (Sabramz) [inline Core/Inc/segment.h:65]: Changed types to work with bms config

File: Core/Inc/stateMachine.h [hunk 5]
```diff
@@ -21,6 +21,7 @@ class StateMachine
         tristate_timer overVolt_tmr;
         tristate_timer lowCell_tmr;
         tristate_timer highTemp_tmr;
+        tristate_timer internalCellComm_tmr;
```
Comment (Sabramz) [inline Core/Inc/stateMachine.h:24]: Timer for error

File: Core/Src/stateMachine.cpp [hunk 6]
```diff
@@ -198,16 +198,17 @@ uint32_t StateMachine::faultReturn(AccumulatorData_t *accData)
...
+            {.id = "High Temp",                  .timer =         highTemp_tmr,   .data_1 =    accData->max_temp.val, .optype_1 = GT,  .lim_1 =                                          MAX_CELL_TEMP, .timeout =         LOW_CELL_TIME, .code =                     PACK_TOO_HOT   /* -----------------------------------UNUSED---------------------------------*/  }, 
+            {.id = "Extremely Low Voltage",      .timer =          lowCell_tmr,   .data_1 = accData->min_voltage.val, .optype_1 = LT,  .lim_1 =                                                    900, .timeout =        HIGH_TEMP_TIME, .code =                 LOW_CELL_VOLTAGE   /* -----------------------------------UNUSED---------------------------------*/  },
+			{.id = "Internal Cell Communication" .timer = internalCellComm_tmr,   .data_1 =   CELL_COMM_FAULT_STATUS, .optype_1 = NEQ, .lim_1 =                                                     0,  .timeout = INTERN_CELL_COMM_TIME, .code =         INTERNAL_CELL_COMM_FAULT   /* -----------------------------------UNUSED---------------------------------*/  },
```
Comment (Sabramz) [inline Core/Src/stateMachine.cpp:211]: Lots of lines changed cuz formatting

File: Core/Src/segment.c (hunk 1 above)
Comment (Sabramz) [reply Core/Src/segment.c:50]: We talked and established that the both errors are internal cell communication faults, so a therm and voltage read error are both covered by the one error code.

File: Core/Src/segment.c [hunk 7]
```diff
@@ -44,6 +45,14 @@ void SegmentInterface::retrieveSegmentData(ChipData_t databuf[NUM_CHIPS])
...
 
     segment_data = nullptr;
+
+    uint16_t err = 0;
+    if (voltage_error) {
+        err = INTERNAL_CELL_COMM_FAULT;
+    }else if (therm_error) {
+        err = INTERNAL_CELL_COMM_FAULT;
+    }
+    CELL_COMM_FAULT_STATUS = err;
```
Comment (Sabramz) [inline Core/Src/segment.c:55]: Passes error status to a shared variable that keeps track of the fault state.

**Comment by dyldonahue** [comment]
Manually Merged from other branch, didnt wanna deal with conflicts

**Pull request closed without merging**